File: guardian-agent-apk/src/com/fleetops/guardian/BootReceiver.java
package com.fleetops.guardian;

import android.content.*;

public class BootReceiver extends BroadcastReceiver {
    @Override
    public void onReceive(Context ctx, Intent intent) {
        if (Intent.ACTION_BOOT_COMPLETED.equals(intent.getAction())) {
            DevicePrefs prefs = new DevicePrefs(ctx);
            if (prefs.isEnrolled()) {
                Intent svc = new Intent(ctx, GuardianService.class);
                if (android.os.Build.VERSION.SDK_INT >= 26) {
                    ctx.startForegroundService(svc);
                } else {
                    ctx.startService(svc);
                }
            }
        }
    }
}

File: apps/guardian-android/app/src/main/kotlin/io/sonalit/guardian/receiver/BootReceiver.kt
package io.sonalit.guardian.receiver

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import io.sonalit.guardian.service.GuardianService
import io.sonalit.guardian.worker.HeartbeatWorker
import io.sonalit.guardian.worker.SyncWorker

class BootReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
        val action = intent.action ?: return
        if (action != Intent.ACTION_BOOT_COMPLETED &&
            action != "android.intent.action.LOCKED_BOOT_COMPLETED" &&
            action != "android.intent.action.MY_PACKAGE_REPLACED"
        ) return

        val prefs = openPrefs(context) ?: return
        val deviceId = prefs.getString("device_id", null) ?: return
        context.startForegroundService(Intent(context, GuardianService::class.java))
        HeartbeatWorker.schedule(context, deviceId = deviceId)
        SyncWorker.schedule(context)
    }

    private fun openPrefs(context: Context): android.content.SharedPreferences? {
        return try {
            val masterKey = MasterKey.Builder(context)
                .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
                .build()
            EncryptedSharedPreferences.create(
                context,
                "guardian_prefs",
                masterKey,
                EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
                EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
            )
        } catch (_: Exception) {
            null
        }
    }
}
